misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://wso2.com) All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.formatter.core;
+
+import io.ballerina.tools.text.LinePosition;
+import io.ballerina.tools.text.LineRange;
+import io.ballerinalang.compiler.syntax.tree.AbstractNodeFactory;
+import io.ballerinalang.compiler.syntax.tree.ChildNodeList;
+import io.ballerinalang.compiler.syntax.tree.Minutiae;
+import io.ballerinalang.compiler.syntax.tree.MinutiaeList;
+import io.ballerinalang.compiler.syntax.tree.Node;
+import io.ballerinalang.compiler.syntax.tree.NonTerminalNode;
+import io.ballerinalang.compiler.syntax.tree.SyntaxKind;
+import io.ballerinalang.compiler.syntax.tree.Token;
+import org.wso2.ballerinalang.compiler.util.diagnotic.DiagnosticPos;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static io.ballerinalang.compiler.syntax.tree.AbstractNodeFactory.createMinutiaeList;
+import static io.ballerinalang.compiler.syntax.tree.AbstractNodeFactory.createWhitespaceMinutiae;
+
+/**
+ * Class that contains the util functions used by the formatting tree modifier.
+ */
+class FormatterUtils {
+
+    private static final String LINE_SEPARATOR = "line.separator";
+
+    /**
+     * Get the node position.
+     *
+     * @param node node
+     * @return node position
+     */
+    static DiagnosticPos getPosition(Node node) {
+        if (node == null) {
+            return null;
+        }
+        LineRange range = node.lineRange();
+        LinePosition startPos = range.startLine();
+        LinePosition endPos = range.endLine();
+        int startOffset = startPos.offset();
+        if (node.kind() == (SyntaxKind.FUNCTION_DEFINITION) || node.kind() == (SyntaxKind.TYPE_DEFINITION) ||
+                node.kind() == (SyntaxKind.CONST_DECLARATION) || node.kind() == (SyntaxKind.OBJECT_TYPE_DESC)) {
+            startOffset = (startOffset / 4) * 4;
+        }
+        return new DiagnosticPos(null, startPos.line() + 1, endPos.line() + 1,
+                startOffset, endPos.offset());
+    }
+
+    // TODO: Use a generic way to get the parent node using querying.
+    static <T extends Node> Node getParent(T node, SyntaxKind syntaxKind) {
```

Comment:
Instead of returning a possibly null node, shall we use Optional here?

Reply:
Since this method is called recursively to find the effective parent node for indentation, converting its return type to optional seems to clutter the code without much of a benefit IMO, or is there something that I might be missing here?

A null return value is expected in a case where a node has no parent. Here, there won't be an addition of columns to the existing indentation.